crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2830,6 +2830,93 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::USub, Type::BooleanLiteral(bool)) => Type::IntLiteral(-i64::from(bool)),
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
+            (UnaryOp::Not, Type::Instance(InstanceType { class })) => {
+                // TODO: move to Type::bool method
+                // I initially wrote this code in the bool method but it caused a cycle.
+                // Also this requires refactoring of that method to pass the node and diagnostics
+                // so we can emit the diagnostic when it's not possible to get the boolean of the
+                // type.
+                let bool_type = if let Symbol::Type(bool_method, _) =
+                    class.class_member(self.db, "__bool__")
+                {
+                    let call = bool_method.call(self.db, &[operand_type]);
+                    let bool_rt = match call.return_ty_result(
+                        self.db,
+                        AnyNodeRef::ExprUnaryOp(unary),
+                        &mut self.diagnostics,
+                    ) {
+                        Ok(t) => t,
+                        Err(_) => Type::BooleanLiteral(true),
+                    };
+
+                    if bool_rt.is_subtype_of(
+                        self.db,
+                        KnownClass::Bool.to_class(self.db).to_instance(self.db),
+                    ) {
+                        bool_rt.bool(self.db)
+                    } else {
+                        self.diagnostics.add(
+                            unary.into(),
+                            "unsupported-operator",
+                            format_args!(
+                                "Method __bool__ for type `{}` should return `bool`, returned type `{}`",
```

Comment:
same as https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2AUNQMaXkDOTUAggFzVQ9QAmJYFAD6wgEZgwlUQAomJSsACUAWgB8mFDC69dUECRgBXECigAGWgYBuJKsPgISMlGHxsZSpUA

Reply:
Yes, I see the pyright precedent, but I think pyright is wrong (or at least sub-optimal) in issuing this diagnostic at the call site rather than at the definition of the class.